Table rules="rows" draws the rule on the cells instead of the rows

karlcow opened the pull request:
#### 30b998b5593acb3e4f5c9ec59b78c172b93aca0d
<pre>
Table rules=&quot;rows&quot; draws the rule on the cells instead of the rows
<a href="https://bugs.webkit.org/show_bug.cgi?id=257254">https://bugs.webkit.org/show_bug.cgi?id=257254</a>
<a href="https://rdar.apple.com/109765480">rdar://109765480</a>

Reviewed by NOBODY (OOPS!).

With rules=&quot;rows&quot; the line between two rows belongs to the row, not to
the cells. WebKit put it on the top and bottom of every cell, so
getComputedStyle on a cell reported a 1px solid border where there is
none.

Fix by leaving the cells without a border and giving the rows a shared
style, the way row groups already get one for rules=&quot;groups&quot;. It uses the
block-direction properties so the rule follows the writing mode. The
table looks the same as before, and as Firefox.

Also invalidate the rows when the attribute changes. The old walk stops
at the first cell it finds, so a table with rows but no cells kept its
previous rule.

Two tests are added. Nothing upstream checked computed styles for these
attributes, and <a href="https://bugs.webkit.org/show_bug.cgi?id=256850">https://bugs.webkit.org/show_bug.cgi?id=256850</a> only had a
reftest. Two subtests fail: rules=cols and rules=groups in a vertical
writing mode still use the physical properties, tracked in
<a href="https://bugs.webkit.org/show_bug.cgi?id=321071">https://bugs.webkit.org/show_bug.cgi?id=321071</a>

* LayoutTests/fast/table/border-changes-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/rendering/non-replaced-elements/tables/table-border-attributes-computed-style-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/html/rendering/non-replaced-elements/tables/table-border-attributes-computed-style.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/html/rendering/non-replaced-elements/tables/table-rules-computed-style-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/html/rendering/non-replaced-elements/tables/table-rules-computed-style.html: Added.
* LayoutTests/platform/glib/fast/table/frame-and-rules-expected.txt:
* LayoutTests/platform/glib/fast/table/rules-attr-dynchange2-expected.txt:
* LayoutTests/platform/glib/tables/mozilla/core/table_rules-expected.txt:
* LayoutTests/platform/glib/tables/mozilla_expected_failures/marvin/table_rules_rows-expected.txt:
* LayoutTests/platform/glib/tables/mozilla_expected_failures/marvin/x_table_rules_rows-expected.txt:
* LayoutTests/platform/ios/fast/table/frame-and-rules-expected.txt:
* LayoutTests/platform/ios/fast/table/rules-attr-dynchange2-expected.txt:
* LayoutTests/platform/ios/tables/mozilla/core/table_rules-expected.txt:
* LayoutTests/platform/ios/tables/mozilla_expected_failures/marvin/table_rules_rows-expected.txt:
* LayoutTests/platform/ios/tables/mozilla_expected_failures/marvin/x_table_rules_rows-expected.txt:
* LayoutTests/platform/mac/fast/table/frame-and-rules-expected.txt:
* LayoutTests/platform/mac/fast/table/rules-attr-dynchange2-expected.txt:
* LayoutTests/platform/mac/tables/mozilla/core/table_rules-expected.txt:
* LayoutTests/platform/mac/tables/mozilla_expected_failures/marvin/table_rules_rows-expected.txt:
* LayoutTests/platform/mac/tables/mozilla_expected_failures/marvin/x_table_rules_rows-expected.txt:
* LayoutTests/platform/win/fast/table/frame-and-rules-expected.txt:
* LayoutTests/platform/win/fast/table/rules-attr-dynchange2-expected.txt:
* LayoutTests/platform/win/tables/mozilla/core/table_rules-expected.txt:
* LayoutTests/platform/win/tables/mozilla_expected_failures/marvin/table_rules_rows-expected.txt:
* LayoutTests/platform/win/tables/mozilla_expected_failures/marvin/x_table_rules_rows-expected.txt:
  The border annotation moves from the cell lines to the row lines.
  Nothing moves or changes size. Every port has its own dump for these
  tests, so all four are updated.

* Source/WebCore/html/HTMLTableElement.cpp:
(WebCore::HTMLTableElement::attributeChanged):
(WebCore::HTMLTableElement::createSharedCellStyle const):
(WebCore::makeRowsRuleStyle):
(WebCore::HTMLTableElement::additionalRowStyle const):
* Source/WebCore/html/HTMLTableElement.h:
* Source/WebCore/html/HTMLTableRowElement.cpp:
(WebCore::HTMLTableRowElement::additionalPresentationalHintStyle const):
* Source/WebCore/html/HTMLTableRowElement.h:
</pre><!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/30b998b5593acb3e4f5c9ec59b78c172b93aca0d

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/179153 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/53092 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/168/builds/46887 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/188984 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/143462 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/39589780-be51-444c-8d83-e959c4b84d2e) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/181016 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/54385 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/52802 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/139703 "Passed tests") | [❌ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/96761 "1 flakes 3 failures") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/05bc6753-c680-4670-bf7b-7e27a963e8e5) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/182110 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/43298 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/160460 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/120150 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/690fa2a9-3a37-4889-b874-6f275d8f9c86) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/41969 "Passed tests") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/40314 "Passed tests") | | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/152369 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/169/builds/39964 "Passed tests") | [✅ 🛠 gtk3-gcc](https://ews-build.webkit.org/#/builders/284/builds/290 "Built successfully") | | 
| | [✅ 🛠 ios-safer-cpp](https://ews-build.webkit.org/#/builders/174/builds/45698 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/44560 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/191202 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/52762 "Built successfully") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/159977 "Build is in progress. Recent messages:OS: Sequoia (15.7.3), Xcode: 26.2; Checked out pull request; Passed layout tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/148942 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/53442 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/170/builds/36590 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/148688 "Passed tests") | | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/43371 "Passed tests") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/125713 "Built successfully") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/120555 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/51960 "Built successfully") | [❌ 🧪 mac-site-isolation](https://ews-build.webkit.org/#/builders/185/builds/14943 "") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/51345 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/51586 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/51406 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/5eb4f6497854e0e166d8b26285991bdf99eb2de5. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/21850

karlcow commented:
If there is no WPT progression, it means probably there are missing WPT tests.

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/2150d9c550594734cdb57b99e5150773cad6552c. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/21850

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/2150d9c550594734cdb57b99e5150773cad6552c. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/21850

karlcow committed:
fix table rules="row" presentational attributes
https://bugs.webkit.org/show_bug.cgi?id=257254
rdar://109765480

Reviewed by NOBODY (OOPS!).

This fixes a table failures with regards to presentation attributes.
It was started with https://bugs.webkit.org/show_bug.cgi?id=256850
This patch finishes it when rules="row" is added to the table. It
aligns the behavior with Firefox.

* LayoutTests/fast/table/border-changes-expected.txt:
* LayoutTests/platform/glib/fast/table/frame-and-rules-expected.txt:
* LayoutTests/platform/glib/fast/table/rules-attr-dynchange2-expected.txt:
* LayoutTests/platform/glib/tables/mozilla_expected_failures/marvin/table_rules_rows-expected.txt:
* LayoutTests/platform/glib/tables/mozilla_expected_failures/marvin/x_table_rules_rows-expected.txt:
* LayoutTests/platform/glib/tables/mozilla/core/table_rules-expected.txt:
* LayoutTests/platform/ios/fast/table/frame-and-rules-expected.txt:
* LayoutTests/platform/ios/fast/table/rules-attr-dynchange2-expected.txt:
* LayoutTests/platform/ios/tables/mozilla_expected_failures/marvin/table_rules_rows-expected.txt:
* LayoutTests/platform/ios/tables/mozilla_expected_failures/marvin/x_table_rules_rows-expected.txt:
* LayoutTests/platform/ios/tables/mozilla/core/table_rules-expected.txt:
* LayoutTests/platform/mac/fast/table/frame-and-rules-expected.txt:
* LayoutTests/platform/mac/fast/table/rules-attr-dynchange2-expected.txt:
* LayoutTests/platform/mac/tables/mozilla_expected_failures/marvin/table_rules_rows-expected.txt:
* LayoutTests/platform/mac/tables/mozilla_expected_failures/marvin/x_table_rules_rows-expected.txt:
* LayoutTests/platform/mac/tables/mozilla/core/table_rules-expected.txt:
* LayoutTests/TestExpectations:
* Source/WebCore/html/HTMLTableElement.cpp:
(WebCore::HTMLTableElement::createSharedCellStyle const):

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/9e5d69ef3a00867595da47e7b11bf910c74299a4. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/21850

nt1m commented on Source/WebCore/html/HTMLTableElement.cpp:
```diff
@@ -509,10 +509,10 @@ Ref<MutableStyleProperties> HTMLTableElement::createSharedCellStyle() const
-        style->setProperty(CSSPropertyBorderBottomWidth, CSSValueThin);
-        style->setProperty(CSSPropertyBorderTopStyle, CSSValueSolid);
-        style->setProperty(CSSPropertyBorderBottomStyle, CSSValueSolid);
+        style->setProperty(CSSPropertyBorderTopWidth, CSSPrimitiveValue::create(0, CSSUnitType::CSS_PX));
+        style->setProperty(CSSPropertyBorderBottomWidth, CSSPrimitiveValue::create(0, CSSUnitType::CSS_PX));
+        style->setProperty(CSSPropertyBorderTopStyle, CSSPrimitiveValue::create(CSSValueNone));
+        style->setProperty(CSSPropertyBorderBottomStyle, CSSPrimitiveValue::create(CSSValueNone));
```
Where in the spec is this defined? Can you provide a link?

I only found: https://html.spec.whatwg.org/#tables-2 but it doesn't quite match what you're defining here.

karlcow replied:
Let me re-check.
https://html.spec.whatwg.org/multipage/rendering.html#tables-2:presentational-hints
And I will create a WPT test for it.
Thanks!

karlcow committed:
Table rules="rows" draws the rule on the cells instead of the rows
https://bugs.webkit.org/show_bug.cgi?id=257254
rdar://109765480

Reviewed by NOBODY (OOPS!).

With rules="rows" the line between two rows belongs to the row, not to
the cells. WebKit put it on the top and bottom of every cell, so
getComputedStyle on a cell reported a 1px solid border where there is
none.

Fix by leaving the cells without a border and giving the rows a shared
style, the way row groups already get one for rules="groups". It uses the
block-direction properties so the rule follows the writing mode. The
table looks the same as before, and as Firefox.

Also invalidate the rows when the attribute changes. The old walk stops
at the first cell it finds, so a table with rows but no cells kept its
previous rule.

Two tests are added. Nothing upstream checked computed styles for these
attributes, and https://bugs.webkit.org/show_bug.cgi?id=256850 only had a
reftest. Two subtests fail: rules=cols and rules=groups in a vertical
writing mode still use the physical properties, tracked in
https://bugs.webkit.org/show_bug.cgi?id=321071

* LayoutTests/fast/table/border-changes-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/rendering/non-replaced-elements/tables/table-border-attributes-computed-style-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/html/rendering/non-replaced-elements/tables/table-border-attributes-computed-style.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/html/rendering/non-replaced-elements/tables/table-rules-computed-style-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/html/rendering/non-replaced-elements/tables/table-rules-computed-style.html: Added.
* LayoutTests/platform/glib/fast/table/frame-and-rules-expected.txt:
* LayoutTests/platform/glib/fast/table/rules-attr-dynchange2-expected.txt:
* LayoutTests/platform/glib/tables/mozilla/core/table_rules-expected.txt:
* LayoutTests/platform/glib/tables/mozilla_expected_failures/marvin/table_rules_rows-expected.txt:
* LayoutTests/platform/glib/tables/mozilla_expected_failures/marvin/x_table_rules_rows-expected.txt:
* LayoutTests/platform/ios/fast/table/frame-and-rules-expected.txt:
* LayoutTests/platform/ios/fast/table/rules-attr-dynchange2-expected.txt:
* LayoutTests/platform/ios/tables/mozilla/core/table_rules-expected.txt:
* LayoutTests/platform/ios/tables/mozilla_expected_failures/marvin/table_rules_rows-expected.txt:
* LayoutTests/platform/ios/tables/mozilla_expected_failures/marvin/x_table_rules_rows-expected.txt:
* LayoutTests/platform/mac/fast/table/frame-and-rules-expected.txt:
* LayoutTests/platform/mac/fast/table/rules-attr-dynchange2-expected.txt:
* LayoutTests/platform/mac/tables/mozilla/core/table_rules-expected.txt:
* LayoutTests/platform/mac/tables/mozilla_expected_failures/marvin/table_rules_rows-expected.txt:
* LayoutTests/platform/mac/tables/mozilla_expected_failures/marvin/x_table_rules_rows-expected.txt:
* LayoutTests/platform/win/fast/table/frame-and-rules-expected.txt:
* LayoutTests/platform/win/fast/table/rules-attr-dynchange2-expected.txt:
* LayoutTests/platform/win/tables/mozilla/core/table_rules-expected.txt:
* LayoutTests/platform/win/tables/mozilla_expected_failures/marvin/table_rules_rows-expected.txt:
* LayoutTests/platform/win/tables/mozilla_expected_failures/marvin/x_table_rules_rows-expected.txt:
  The border annotation moves from the cell lines to the row lines.
  Nothing moves or changes size. Every port has its own dump for these
  tests, so all four are updated.

* Source/WebCore/html/HTMLTableElement.cpp:
(WebCore::HTMLTableElement::attributeChanged):
(WebCore::HTMLTableElement::createSharedCellStyle const):
(WebCore::makeRowsRuleStyle):
(WebCore::HTMLTableElement::additionalRowStyle const):
* Source/WebCore/html/HTMLTableElement.h:
* Source/WebCore/html/HTMLTableRowElement.cpp:
(WebCore::HTMLTableRowElement::additionalPresentationalHintStyle const):
* Source/WebCore/html/HTMLTableRowElement.h:
